lib/services/access_request_test.go

```diff
@@ -1956,6 +2012,128 @@ func TestPruneMappedRoles(t *testing.T) {
 	}
 }
 
+func TestValidateSAMLIdPAccessRequest(t *testing.T) {
+	ctx := context.Background()
+	g, username := newFixture(t)
+
+	uls, ok := g.userStates[username]
+	require.True(t, ok, "user should exist")
+
+	testCases := []struct {
+		desc               string
+		loginHint          string
+		requestResourceIDs []types.ResourceID
+		errorAssertion     require.ErrorAssertionFunc
+		overrideRoles      []string
+		expectRoles        []string
+	}{
+		{
+			desc: "has direct access to resource",
+			requestResourceIDs: []types.ResourceID{
+				{
+					ClusterName: g.clusterName,
+					Kind:        types.KindSAMLIdPServiceProvider,
+					Name:        "samlapp1",
+				},
+			},
+			expectRoles:    []string{"app-admins"},
+			errorAssertion: require.NoError,
+		},
+		{
+			desc: "requesting with v7 role fails",
+			requestResourceIDs: []types.ResourceID{
+				{
+					ClusterName: g.clusterName,
+					Kind:        types.KindSAMLIdPServiceProvider,
+					Name:        "samlapp1",
+				},
+			},
+			overrideRoles: []string{"saml-rolev7-requester"},
+			errorAssertion: func(t require.TestingT, err error, args ...interface{}) {
+				require.ErrorContains(t, err, "user may already have access")
+			},
+		},
+		{
+			desc: "requesting with v8 role succeeds",
+			requestResourceIDs: []types.ResourceID{
+				{
+					ClusterName: g.clusterName,
+					Kind:        types.KindSAMLIdPServiceProvider,
+					Name:        "samlapp1",
+				},
+			},
+			overrideRoles:  []string{"response-team"},
+			errorAssertion: require.NoError,
+		},
+		{
+			desc: "v7 role grants access, but can't search resource using v8 role",
+			requestResourceIDs: []types.ResourceID{
+				{
+					ClusterName: g.clusterName,
+					Kind:        types.KindSAMLIdPServiceProvider,
+					Name:        "samlapp1",
+				},
+			},
+			expectRoles:   []string{"app-admins"},
+			overrideRoles: []string{"saml-rolev7", "saml-v8-dev-requester"},
+			errorAssertion: func(t require.TestingT, err error, args ...interface{}) {
+				require.ErrorContains(t, err, "no roles configured")
+			},
+		},
+		{
+			desc: "request fails if v8 role exists and does not grant access",
```

Comment:
can we have a testcase where it's possible to request 2 v8 roles, one which grants access to the app and one which does not, and make sure it requests the correct role? I don't think the code you have right now will work for this case, but I think it should